// eslint-plugin-better-tailwindcss/lib/esm/rules/no-conflicting-classes.js
import { DEFAULT_ATTRIBUTE_NAMES, DEFAULT_CALLEE_NAMES, DEFAULT_TAG_NAMES, DEFAULT_VARIABLE_NAMES } from "../options/default-options.js";
import { ATTRIBUTE_SCHEMA, CALLEE_SCHEMA, ENTRYPOINT_SCHEMA, TAG_SCHEMA, TAILWIND_CONFIG_SCHEMA, VARIABLE_SCHEMA } from "../options/descriptions.js";
import { getConflictingClasses } from "../tailwindcss/conflicting-classes.js";
import { getCommonOptions } from "../utils/options.js";
import { createRuleListener } from "../utils/rule.js";
import { augmentMessageWithWarnings, getExactClassLocation, splitClasses } from "../utils/utils.js";
const defaultOptions = {
    attributes: DEFAULT_ATTRIBUTE_NAMES,
    callees: DEFAULT_CALLEE_NAMES,
    tags: DEFAULT_TAG_NAMES,
    variables: DEFAULT_VARIABLE_NAMES
};
const DOCUMENTATION_URL = "https://github.com/schoero/eslint-plugin-better-tailwindcss/blob/main/docs/rules/no-conflicting-classes.md";
export const noConflictingClasses = {
    name: "no-conflicting-classes",
    rule: {
        create: ctx => createRuleListener(ctx, getOptions(ctx), lintLiterals),
        meta: {
            docs: {
                description: "Disallow classes that produce conflicting styles.",
                recommended: false,
                url: DOCUMENTATION_URL
            },
            fixable: "code",
            schema: [
                {
                    additionalProperties: false,
                    properties: {
                        ...CALLEE_SCHEMA,
                        ...ATTRIBUTE_SCHEMA,
                        ...VARIABLE_SCHEMA,
                        ...TAG_SCHEMA,
                        ...ENTRYPOINT_SCHEMA,
                        ...TAILWIND_CONFIG_SCHEMA
                    },
                    type: "object"
                }
            ],
            type: "problem"
        }
    }
};
function lintLiterals(ctx, literals) {
    for (const literal of literals) {
        const { tailwindConfig } = getOptions(ctx);
        const classes = splitClasses(literal.content);
        const { conflictingClasses, warnings } = getConflictingClasses({ classes, configPath: tailwindConfig, cwd: ctx.cwd });
        if (Object.keys(conflictingClasses).length === 0) {
            continue;
        }
        for (const className in conflictingClasses) {
            const conflicts = Object.entries(conflictingClasses[className]);
            if (conflicts.length === 0) {
                continue;
            }
            const conflictingClassNames = conflicts.map(([conflictingClassName]) => conflictingClassName);
            const conflictingProperties = conflicts.reduce((acc, [, properties]) => {
                for (const property of properties) {
                    if (!acc.includes(property.cssPropertyName)) {
                        acc.push(property.cssPropertyName);
                    }
                }
                return acc;
            }, []);
            ctx.report({
                data: {
                    className,
                    conflictingClassNames: conflictingClassNames.join(", "),
                    conflictingProperties: conflictingProperties.map(conflictingProperty => `"${conflictingProperty}"`).join(", ")
                },
                loc: getExactClassLocation(literal, className),
                message: augmentMessageWithWarnings("Conflicting class detected: \"{{ className }}\" and \"{{ conflictingClassNames }}\" apply the same CSS properties: {{ conflictingProperties }}.", DOCUMENTATION_URL, warnings)
            });
        }
    }
}
export function getOptions(ctx) {
    return getCommonOptions(ctx);
}
