// eslint-plugin-better-tailwindcss/lib/esm/async-utils/version.js
import { readFileSync } from "node:fs";
import { withCache } from "./cache.js";
import { jsonResolver } from "./resolvers.js";
export var TailwindcssVersion;
(function (TailwindcssVersion) {
    TailwindcssVersion[TailwindcssVersion["V3"] = 3] = "V3";
    TailwindcssVersion[TailwindcssVersion["V4"] = 4] = "V4";
})(TailwindcssVersion || (TailwindcssVersion = {}));
export function isSupportedVersion(version) {
    return version === TailwindcssVersion.V3 || version === TailwindcssVersion.V4;
}
export function isTailwindcssVersion3(version) {
    return version === TailwindcssVersion.V3;
}
export function isTailwindcssVersion4(version) {
    return version === TailwindcssVersion.V4;
}
export function getTailwindcssVersion() {
    const packageJsonPath = jsonResolver.resolveSync({}, process.cwd(), "tailwindcss/package.json");
    if (!packageJsonPath) {
        throw new Error("Could not find a Tailwind CSS package.json");
    }
    return withCache(packageJsonPath, () => {
        const packageJson = JSON.parse(readFileSync(packageJsonPath, "utf-8"));
        if (!packageJson) {
            throw new Error("Error reading Tailwind CSS package.json");
        }
        return parseSemanticVersion(packageJson.version);
    });
}
function parseSemanticVersion(version) {
    const [major, minor, patchString] = version.split(".");
    const [patch, identifier] = patchString.split("-");
    return { identifier, major: +major, minor: +minor, patch: +patch };
}
