// eslint-plugin-better-tailwindcss/lib/cjs/rules/no-unregistered-classes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noUnregisteredClasses = exports.DEFAULT_IGNORED_UNREGISTERED_CLASSES = void 0;
exports.getOptions = getOptions;
const default_options_js_1 = require("../options/default-options.js");
const descriptions_js_1 = require("../options/descriptions.js");
const custom_component_classes_js_1 = require("../tailwindcss/custom-component-classes.js");
const prefix_js_1 = require("../tailwindcss/prefix.js");
const unregistered_classes_js_1 = require("../tailwindcss/unregistered-classes.js");
const options_js_1 = require("../utils/options.js");
const rule_js_1 = require("../utils/rule.js");
const utils_js_1 = require("../utils/utils.js");
exports.DEFAULT_IGNORED_UNREGISTERED_CLASSES = [];
const defaultOptions = {
    attributes: default_options_js_1.DEFAULT_ATTRIBUTE_NAMES,
    callees: default_options_js_1.DEFAULT_CALLEE_NAMES,
    detectComponentClasses: false,
    ignore: exports.DEFAULT_IGNORED_UNREGISTERED_CLASSES,
    tags: default_options_js_1.DEFAULT_TAG_NAMES,
    variables: default_options_js_1.DEFAULT_VARIABLE_NAMES
};
const DOCUMENTATION_URL = "https://github.com/schoero/eslint-plugin-better-tailwindcss/blob/main/docs/rules/no-unregistered-classes.md";
exports.noUnregisteredClasses = {
    name: "no-unregistered-classes",
    rule: {
        create: ctx => (0, rule_js_1.createRuleListener)(ctx, getOptions(ctx), lintLiterals),
        meta: {
            docs: {
                description: "Disallow any css classes that are not registered in tailwindcss.",
                recommended: false,
                url: DOCUMENTATION_URL
            },
            fixable: "code",
            schema: [
                {
                    additionalProperties: false,
                    properties: {
                        ...descriptions_js_1.CALLEE_SCHEMA,
                        ...descriptions_js_1.ATTRIBUTE_SCHEMA,
                        ...descriptions_js_1.VARIABLE_SCHEMA,
                        ...descriptions_js_1.TAG_SCHEMA,
                        ...descriptions_js_1.ENTRYPOINT_SCHEMA,
                        ...descriptions_js_1.TAILWIND_CONFIG_SCHEMA,
                        detectComponentClasses: {
                            default: defaultOptions.detectComponentClasses,
                            description: "Whether to automatically detect custom component classes from the tailwindcss config.",
                            type: "boolean"
                        },
                        ignore: {
                            description: "A list of classes that should be ignored by the rule.",
                            items: {
                                type: "string"
                            },
                            type: "array"
                        }
                    },
                    type: "object"
                }
            ],
            type: "problem"
        }
    }
};
function lintLiterals(ctx, literals) {
    const { detectComponentClasses, ignore, tailwindConfig } = getOptions(ctx);
    const { prefix, suffix } = (0, prefix_js_1.getPrefix)({ configPath: tailwindConfig, cwd: ctx.cwd });
    const ignoredGroups = new RegExp(`^${(0, utils_js_1.escapeForRegex)(`${prefix}${suffix}`)}group(?:\\/(\\S*))?$`);
    const ignoredPeers = new RegExp(`^${(0, utils_js_1.escapeForRegex)(`${prefix}${suffix}`)}peer(?:\\/(\\S*))?$`);
    const customComponentClasses = detectComponentClasses
        ? (0, custom_component_classes_js_1.getCustomComponentClasses)({ configPath: tailwindConfig, cwd: ctx.cwd })
        : [];
    for (const literal of literals) {
        const classes = (0, utils_js_1.splitClasses)(literal.content);
        const { unregisteredClasses, warnings } = (0, unregistered_classes_js_1.getUnregisteredClasses)({ classes, configPath: tailwindConfig, cwd: ctx.cwd });
        if (unregisteredClasses.length === 0) {
            continue;
        }
        for (const unregisteredClass of unregisteredClasses) {
            if (ignore.some(ignoredClass => unregisteredClass.match(ignoredClass)) ||
                customComponentClasses.includes(unregisteredClass) ||
                unregisteredClass.match(ignoredGroups) ||
                unregisteredClass.match(ignoredPeers)) {
                continue;
            }
            ctx.report({
                data: {
                    unregistered: (0, utils_js_1.display)(unregisteredClass)
                },
                loc: (0, utils_js_1.getExactClassLocation)(literal, unregisteredClass),
                message: (0, utils_js_1.augmentMessageWithWarnings)("Unregistered class detected: {{ unregistered }}", DOCUMENTATION_URL, warnings)
            });
        }
    }
}
function getOptions(ctx) {
    const options = ctx.options[0] ?? {};
    const common = (0, options_js_1.getCommonOptions)(ctx);
    const ignore = options.ignore ?? defaultOptions.ignore;
    const detectComponentClasses = options.detectComponentClasses ?? defaultOptions.detectComponentClasses;
    return {
        ...common,
        detectComponentClasses,
        ignore
    };
}
